SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp

```diff
@@ -137,45 +137,67 @@ void USpatialReceiver::OnAddComponent(Worker_AddComponentOp& Op)
 		return;
 	}
 
-	// If a client gains ownership over something it had already checked out, it will
-	// add component interest on the owner only data components, which will trigger an
-	// AddComponentOp, but it is not guaranteed to be inside a critical section.
-	if (!NetDriver->IsServer())
+	if (bInCriticalSection)
 	{
-		if (USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(Op.entity_id))
-		{
-			if (ClassInfoManager->GetCategoryByComponentId(Op.data.component_id) == SCHEMA_OwnerOnly)
-			{
-				// We received owner only data, and we have the entity checked out already,
-				// so this happened as a result of adding component interest. Apply the data
-				// immediately instead of queuing it up (since there will be no AddEntityOp).
-				ApplyComponentData(Op.entity_id, Op.data, Channel);
-				return;
-			}
-		}
+		PendingAddComponents.Emplace(Op.entity_id, Op.data.component_id, MakeUnique<DynamicComponent>(Op.data));
 	}
-
-	if (!bInCriticalSection)
+	else
 	{
-		UE_LOG(LogSpatialReceiver, Warning, TEXT("Received a dynamically added component, these are currently unsupported - component ID: %u entity ID: %lld"),
-			Op.data.component_id, Op.entity_id);
-		return;
+		HandleDynamicAddComponent(Op);
 	}
-	PendingAddComponents.Emplace(Op.entity_id, Op.data.component_id, MakeUnique<DynamicComponent>(Op.data));
 }
 
-void USpatialReceiver::OnRemoveEntity(Worker_RemoveEntityOp& Op)
+void USpatialReceiver::OnRemoveEntity(const Worker_RemoveEntityOp& Op)
 {
 	RemoveActor(Op.entity_id);
 }
 
+void USpatialReceiver::OnRemoveComponent(const Worker_RemoveComponentOp& Op)
+{
+	// If we're in a critical section, we've received a RemoveEntityOp.
+	// Our RemoveEntityOp processing relies on component data which
+	// SpatialOS has told us to remove.
+	// Because of this, we skip over RemoveComponentOps in critical sections
+	// and let the RemoveEntityOp handle cleanup.
+	if (bInCriticalSection)
+	{
+		return;
+	}
+
+	// If we are delegated authority over a component that does not exist, we will receive
+	// a RemoveComponentOp. We should only remove local data if this component does exist.
+	if (!StaticComponentView->HasComponent(Op.entity_id, Op.component_id))
+	{
+		return;
+	}
+
+	if (AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(Op.entity_id).Get()))
+	{
+		if (UObject* Object = PackageMap->GetObjectFromUnrealObjectRef(FUnrealObjectRef(Op.entity_id, Op.component_id)).Get())
+		{
+			USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(Op.entity_id);
+			check(Channel != nullptr);
```

Comment:
Remove. Torn off actors won't have an actor channel but might receive this op.

Reply:
Resolved offline